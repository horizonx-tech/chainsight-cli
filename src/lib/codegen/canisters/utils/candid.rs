use anyhow::Ok;
use candid::{bindings::rust::Target, check_prog, types::Type, IDLProg, TypeEnv};

// NOTE/TODO: Duplicates struct in common.rs, Use this
pub struct CanisterMethodIdentifier {
    pub identifier: String,
    type_env: TypeEnv,
}
impl CanisterMethodIdentifier {
    pub const REQUEST_ARGS_TYPE_NAME: &'static str = "RequestArgsType";
    pub const RESPONSE_TYPE_NAME: &'static str = "ResponseType";

    pub fn new(s: &str) -> anyhow::Result<Self> {
        let (identifier, args_ty, response_ty) = extract_elements(s)?;
        let did: String = Self::generate_did(&args_ty, &response_ty);

        let ast: IDLProg = did.to_string().parse().unwrap();
        let mut type_env = TypeEnv::new();
        let _ = check_prog(&mut type_env, &ast);

        Ok(Self {
            identifier,
            type_env,
        })
    }

    pub fn compile(&self) -> String {
        let mut config = candid::bindings::rust::Config::new();
        // Update the structure derive to chainsight's own settings
        config.type_attributes = "#[derive(Clone, Debug, candid :: CandidType, candid :: Deserialize, serde :: Serialize, chainsight_cdk_macros :: StableMemoryStorable)]".to_string();
        config.target = Target::CanisterStub;
        let contents = candid::bindings::rust::compile(&config, &self.type_env, &None);

        let lines = contents
            .lines()
            // Delete comment lines and blank lines
            .filter(|line| !(line.starts_with("//") || line.is_empty()))
            .collect::<Vec<_>>();
        lines.join("\n")
    }

    pub fn get_types(&self) -> (Option<&Type>, Option<&Type>) {
        (
            self.find_type(Self::REQUEST_ARGS_TYPE_NAME),
            self.find_type(Self::RESPONSE_TYPE_NAME),
        )
    }

    fn generate_did(args_ty: &str, response_ty: &str) -> String {
        let args_ty_did = if args_ty.is_empty() {
            "".to_string()
        } else {
            generate_did_type(Self::REQUEST_ARGS_TYPE_NAME, args_ty)
        };
        let response_ty_did = if response_ty.is_empty() {
            "".to_string()
        } else {
            generate_did_type(Self::RESPONSE_TYPE_NAME, response_ty)
        };
        format!("{}\n{}", args_ty_did, response_ty_did)
    }

    fn find_type(&self, key: &str) -> Option<&Type> {
        let ty = self.type_env.find_type(key);
        ty.ok()
    }
}

fn generate_did_type(key: &str, value: &str) -> String {
    format!("type {} = {};", key, value)
}

fn extract_elements(s: &str) -> anyhow::Result<(String, String, String)> {
    let (identifier, remains) = s
        .split_once(':')
        .expect("Invalid canister method identifier");
    let (args_ty, response_ty) = remains
        .split_once("->")
        .expect("Invalid canister method identifier");

    let trim_type_str = |s: &str| {
        let trimed = s.trim();
        let removed_brackets = trimed.trim_matches(|c| c == '(' || c == ')');
        removed_brackets.trim().to_string()
    };

    Ok((
        identifier.trim().to_string(),
        trim_type_str(args_ty),
        trim_type_str(response_ty),
    ))
}

#[cfg(test)]
mod tests {
    use super::*;

    use insta::assert_snapshot;

    const TEST_IDENTS: &'static [(&'static str, &'static str); 4] = &[
        (&"normal", &"update_value : (nat64) -> (text)"),
        (&"no_args", &"get_value : () -> (text)"),
        (
            &"record",
            &"get_snapshot : (nat64) -> (record { value : text; timestamp : nat64 })",
        ),
        (
            &"nested_record",
            &"get_last_snapshot : () -> (record { dai : record { usd : float64 } })",
        ),
    ];

    #[test]
    fn test_compile() {
        for (label, s) in TEST_IDENTS {
            let ident = CanisterMethodIdentifier::new(s).expect("Failed to parse");
            let compiled = ident.compile();
            assert_snapshot!(format!("compile__{}", label.to_string()), compiled);
        }
    }

    #[test]
    fn test_generate_did() {
        let key = "Account";
        let value = "record { owner : principal; subaccount : opt blob }";
        let did = generate_did_type(key, value);
        assert_snapshot!(did);
    }

    #[test]
    fn test_extract_elements() {
        for (label, s) in TEST_IDENTS {
            let result = extract_elements(s).expect("Failed to parse");
            assert_snapshot!(
                format!("extract_elements__{}", label.to_string()),
                format!("{:#?}", result)
            );
        }
    }

    #[test]
    fn test_extract_elements_multi_nested_record() {
        let s =
            "icrc2_allowance : (record { account : record { owner : principal; subaccount : opt blob }; spender : record { owner : principal; subaccount : opt blob } }) -> (record { allowance : nat; expires_at : opt nat64 })";
        let (identifier, args_ty, response_ty) = extract_elements(s).expect("Failed to parse");
        assert_eq!(identifier, "icrc2_allowance");
        assert_eq!(args_ty, "record { account : record { owner : principal; subaccount : opt blob }; spender : record { owner : principal; subaccount : opt blob } }");
        assert_eq!(
            response_ty,
            "record { allowance : nat; expires_at : opt nat64 }"
        );
    }
}
