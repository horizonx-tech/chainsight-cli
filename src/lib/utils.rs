use std::path::Path;

pub const CHAINSIGHT_FILENAME: &str = ".chainsight";
pub const PROJECT_MANIFEST_FILENAME: &str = "project.yaml";
pub const PROJECT_MANIFEST_VERSION: &str = "v1";

/// Check if .chainsight file exists in project folder
pub fn is_chainsight_project(path: Option<String>) -> Result<(), String> {
    if let Some(path) = path {
        let path = Path::new(&path);
        if !path.exists() {
            return Err(format!("Provided path '{}' does not exist.", path.display()));
        }
        // Check if .chainsight file exists in the provided path
        if !path.join(CHAINSIGHT_FILENAME).exists() {
            return Err(format!("No .chainsight file found in the provided path '{}'.", path.display()));
        }
    } else {
        if !Path::new(CHAINSIGHT_FILENAME).exists() {
            return Err("No .chainsight file found in the current directory.".to_string());
        }
    }
    Ok(())
}
