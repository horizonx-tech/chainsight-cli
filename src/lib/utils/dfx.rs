use std::{ffi::OsStr, fmt, path::Path, str::FromStr};

use super::remove_trailing_newline;

#[derive(Clone, Debug)]
pub enum DfxWrapperNetwork {
    IC,
    Local(Option<u64>),
    Custom(String),
}
impl DfxWrapperNetwork {
    pub fn value(&self) -> String {
        match self {
            DfxWrapperNetwork::IC => "https://ic0.app/".to_string(),
            DfxWrapperNetwork::Local(port) => format!("http://localhost:{}", port.unwrap_or(4943)),
            DfxWrapperNetwork::Custom(custom) => custom.clone(),
        }
    }

    pub fn args(&self) -> Vec<String> {
        let value = match self {
            DfxWrapperNetwork::IC => "ic".to_string(),
            _ => self.value(),
        };
        vec!["--network".to_string(), value]
    }
}
impl Default for DfxWrapperNetwork {
    fn default() -> Self {
        DfxWrapperNetwork::IC
    }
}

#[derive(Clone, Debug, Default)]
pub struct DfxWrapper {
    execution_dir_str: Option<String>,
    network: DfxWrapperNetwork,
}

#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct Version {
    major: u32,
    minor: u32,
    patch: u32,
    pre_release: Option<String>,
}
impl FromStr for Version {
    type Err = &'static str;

    fn from_str(version: &str) -> Result<Self, Self::Err> {
        let parts: Vec<&str> = version.split('-').collect();
        let version_parts: Vec<&str> = parts[0].split('.').collect();

        if version_parts.len() != 3 {
            return Err("Invalid version string");
        }

        let major = version_parts[0]
            .parse::<u32>()
            .map_err(|_| "Invalid major version")?;
        let minor = version_parts[1]
            .parse::<u32>()
            .map_err(|_| "Invalid minor version")?;
        let patch = version_parts[2]
            .parse::<u32>()
            .map_err(|_| "Invalid patch version")?;

        Ok(Version {
            major,
            minor,
            patch,
            pre_release: parts.get(1).map(ToString::to_string),
        })
    }
}
impl fmt::Display for Version {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self.pre_release.as_ref() {
            Some(pr) => write!(f, "{}.{}.{}-{}", self.major, self.minor, self.patch, pr),
            None => write!(f, "{}.{}.{}", self.major, self.minor, self.patch),
        }
    }
}

impl DfxWrapper {
    fn execution_dir(&self) -> &Path {
        let dir = self.execution_dir_str.as_deref().unwrap_or(".");
        Path::new(dir)
    }

    fn args(&self, command: Vec<String>) -> Vec<String> {
        let mut args = vec![command];
        args.push(self.network.args());
        args.concat()
    }

    pub fn new(network: DfxWrapperNetwork, path: Option<String>) -> Self {
        DfxWrapper {
            execution_dir_str: path,
            network,
        }
    }

    pub fn version(&self) -> Result<Version, String> {
        let res = exec_cmd_string_output("dfx", self.execution_dir(), vec!["--version"])
            .map(remove_trailing_newline)?;
        let version_str = res.replace("dfx ", "");
        Version::from_str(&version_str).map_err(|e| e.to_string())
    }

    pub fn identity_whoami(&self) -> Result<String, String> {
        exec_cmd_string_output("dfx", self.execution_dir(), vec!["identity", "whoami"])
            .map(remove_trailing_newline)
    }

    pub fn identity_get_wallet(&self) -> Result<String, String> {
        exec_cmd_string_output(
            "dfx",
            self.execution_dir(),
            self.args(vec!["identity".to_string(), "get-wallet".to_string()]),
        )
        .map(remove_trailing_newline)
    }
}

fn exec_cmd_none_output<I, S>(cmd: &str, execution_dir: &Path, args: I) -> Result<(), String>
where
    I: IntoIterator<Item = S>,
    S: AsRef<OsStr>,
{
    exec_cmd_generic_output(cmd, execution_dir, args, |_stdout| Ok(()))
}

fn exec_cmd_string_output<I, S>(cmd: &str, execution_dir: &Path, args: I) -> Result<String, String>
where
    I: IntoIterator<Item = S>,
    S: AsRef<OsStr>,
{
    exec_cmd_generic_output(cmd, execution_dir, args, |stdout| {
        Ok(std::str::from_utf8(&stdout).unwrap().to_string())
    })
}

fn exec_cmd_json_output<T, I, S>(cmd: &str, execution_dir: &Path, args: I) -> Result<T, String>
where
    T: serde::de::DeserializeOwned,
    I: IntoIterator<Item = S>,
    S: AsRef<OsStr>,
{
    exec_cmd_generic_output(cmd, execution_dir, args, |stdout| {
        Ok(serde_json::from_slice(&stdout).unwrap())
    })
}

fn exec_cmd_generic_output<T, F, I, S>(
    cmd: &str,
    execution_dir: &Path,
    args: I,
    process_output: F,
) -> Result<T, String>
where
    F: FnOnce(Vec<u8>) -> Result<T, String>,
    I: IntoIterator<Item = S>,
    S: AsRef<OsStr>,
{
    let output = exec_cmd(cmd, execution_dir, args)
        .unwrap_or_else(|_| panic!("failed to execute process: {}", cmd));
    if output.status.success() {
        process_output(output.stdout)
    } else {
        Err(std::str::from_utf8(&output.stderr).unwrap().to_string())
    }
}

fn exec_cmd<I, S>(cmd: &str, execution_dir: &Path, args: I) -> std::io::Result<std::process::Output>
where
    I: IntoIterator<Item = S>,
    S: AsRef<OsStr>,
{
    std::process::Command::new(cmd)
        .current_dir(execution_dir)
        .args(args)
        // .stdout(Stdio::piped())
        .output()
}
