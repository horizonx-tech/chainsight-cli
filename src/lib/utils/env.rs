use std::collections::HashMap;

pub fn cache_envfile(env_file_path: Option<&str>) -> anyhow::Result<()> {
    if let Some(env_file_path) = env_file_path {
        dotenvy::from_filename(env_file_path)?;
    } else {
        dotenvy::dotenv()?;
    }
    Ok(())
}

/// replace ${ENV_VAR} with actual value
pub fn load_env(contents: &str) -> anyhow::Result<String> {
    let mut envs = HashMap::new();
    for (k, v) in dotenvy::vars() {
        envs.insert(k, v);
    }
    let mut contents = contents.to_string();
    for (k, v) in envs {
        contents = contents.replace(&format!("${{{}}}", k), &v);
    }
    Ok(contents)
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_load_env() {
        let dotenv_file = r#"
        TEST_ENV=TEST
        TEST_ENV2=TEST2
        "#;
        let contents = r#"
        TEST_ENV: ${TEST_ENV}
        TEST_ENV2: ${TEST_ENV2}
        test3: ${TEST_ENV}
        raw: raw
        "#;
        let expected = r#"
        TEST_ENV: TEST
        TEST_ENV2: TEST2
        test3: TEST
        raw: raw
        "#;
        // setup
        let test_dotenv = "test_dotenv";
        std::fs::write(test_dotenv, dotenv_file).unwrap();
        dotenvy::from_filename(test_dotenv).ok();
        // test
        let actual = load_env(contents).unwrap();
        assert_eq!(actual, expected);
        // teardown
        std::fs::remove_file(test_dotenv).unwrap();
        dotenvy::dotenv().ok();
    }
    #[test]
    fn test_load_env_without_env_file() {
        let contents = r#"
        TEST_ENV: ${TEST_ENV}
        TEST_ENV2: ${TEST_ENV2}
        test3: ${TEST_ENV}
        raw: raw
        "#;
        let expected = contents;
        let actual = load_env(contents).unwrap();
        assert_eq!(actual, expected);
        dotenvy::dotenv().ok();
    }
}
