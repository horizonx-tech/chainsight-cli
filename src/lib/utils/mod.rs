use std::{panic, path::Path};

use inflector::cases::snakecase::to_snake_case;

pub mod clap;
pub mod env;
pub mod interaction;
pub mod paths;
pub mod serializer;

pub const CHAINSIGHT_FILENAME: &str = ".chainsight";
pub const PROJECT_MANIFEST_FILENAME: &str = "project.yaml";
pub const PROJECT_MANIFEST_VERSION: &str = "v1";
pub const DOTENV_FILENAME: &str = ".env";
pub const GITIGNORE_FILENAME: &str = ".gitignore";
pub const ARTIFACTS_DIR: &str = "artifacts";

/// To handle 256bits Unsigned Integer type in ic_web3_rs
pub const U256_TYPE: &str = "ic_web3_rs::types::U256";
/// To handle Address type in ic_web3_rs
pub const ADDRESS_TYPE: &str = "ic_web3_rs::types::Address";

/// Check if .chainsight file exists in project folder
pub fn is_chainsight_project(path: Option<String>) -> Result<(), String> {
    let path_str = path.unwrap_or(".".to_string());
    let path = Path::new(&path_str);
    // Check if .chainsight file exists in the provided or default path
    if !path.join(CHAINSIGHT_FILENAME).exists() {
        return Err(format!(
            "'{}' is not a chainsight project root (no '{}' file exists).",
            path.display(),
            CHAINSIGHT_FILENAME
        ));
    }
    Ok(())
}

/// Convert camelCase String to snake_case
pub fn convert_camel_to_snake(val: &str) -> String {
    // NOTE: use Inflator in ic-solidity-bindgen
    // https://github.com/horizonx-tech/ic-solidity-bindgen/blob/0972bede5957927bcb8f675decd93878b849dc76/ic-solidity-bindgen-macros/src/abi_gen.rs#L192
    to_snake_case(val)
}

/// Outputs duplicate values for a given set of elements.
pub fn find_duplicates<T: Eq + std::hash::Hash>(values: &[T]) -> Vec<&T> {
    let mut duplicates = Vec::new();
    let mut set = std::collections::HashSet::new();
    for value in values {
        if !set.insert(value) {
            duplicates.push(value);
        }
    }
    duplicates
}

pub fn catch_unwind_silent<F: FnOnce() -> R + panic::UnwindSafe, R>(
    f: F,
) -> std::thread::Result<R> {
    let prev_hook = panic::take_hook();
    panic::set_hook(Box::new(|_| {}));
    let result = panic::catch_unwind(f);
    panic::set_hook(prev_hook);
    result
}
