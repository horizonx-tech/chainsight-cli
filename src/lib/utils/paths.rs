pub fn src_path_str(root: &str) -> String {
    format!("{}/src", root)
}

pub fn logics_path_str(src: &str, component: &str) -> String {
    format!("{}/logics/{}", src, component)
}

pub fn canisters_path_str(src: &str, component: &str) -> String {
    format!("{}/canisters/{}", src, component)
}

pub fn bindings_path_str(src: &str, component: &str) -> String {
    format!("{}/bindings/{}", src, bindings_name(component))
}

pub fn canister_name(component: &str) -> String {
    format!("{}_canister", component)
}

pub fn bindings_name(component: &str) -> String {
    format!("{}_bindings", component)
}

pub fn logic_dependency(component: &str) -> String {
    format!(
        r#"{} = {{ path = "../../logics/{}" }}"#,
        component, component
    )
}

pub fn bindings_dependency(component: &str) -> String {
    format!(
        r#"{} = {{ path = "../../bindings/{}" }}"#,
        bindings_name(component),
        bindings_name(component)
    )
}
