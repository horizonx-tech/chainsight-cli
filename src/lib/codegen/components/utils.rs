use chainsight_cdk::convert::candid::CanisterMethodIdentifier;

// Generate types.rs code using the type information in bindings
pub fn generate_types_from_bindings(id: &str, identifier: &str) -> anyhow::Result<String> {
    let identifier = CanisterMethodIdentifier::new(identifier)?;
    let (args_ty, _) = identifier.get_types();

    let mut codes = format!(
        r#"use {}_bindings as bindings;
pub type {} = bindings::{};
"#,
        id,
        CanisterMethodIdentifier::RESPONSE_TYPE_NAME,
        CanisterMethodIdentifier::RESPONSE_TYPE_NAME
    );
    if args_ty.is_some() {
        codes += &format!(
            r#"pub type {} = bindings::{};"#,
            CanisterMethodIdentifier::REQUEST_ARGS_TYPE_NAME,
            CanisterMethodIdentifier::REQUEST_ARGS_TYPE_NAME
        );
    }

    Ok(codes)
}

// determine if the caller is a lens with arguments by CanisterMethodIdentifier
// NOTE: only for snapshot_indexer_icp, relayer
pub fn is_lens_with_args(identifier: CanisterMethodIdentifier) -> bool {
    let (req_ty, _) = identifier.get_types();
    if let Some(req_ty) = req_ty {
        req_ty.to_string() == "LensArgs" // TODO: with key 'LensArgs', necessary to determine if the only type that can be obtained is targets (Vec<String>).
    } else {
        false
    }
}
