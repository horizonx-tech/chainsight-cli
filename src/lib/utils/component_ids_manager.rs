use std::collections::BTreeMap;

use super::dfx::DfxWrapperNetwork;

pub type ComponentIds = BTreeMap<String, String>;
pub struct ComponentIdsManager {
    filename: String,
    components: ComponentIds,
}

impl ComponentIdsManager {
    pub fn new(network: &DfxWrapperNetwork) -> Self {
        Self {
            filename: Self::filename(network),
            components: BTreeMap::new(),
        }
    }

    pub fn load(network: &DfxWrapperNetwork, dir_path: &str) -> anyhow::Result<Self> {
        let filename = Self::filename(network);
        let path = format!("{}/{}", dir_path, filename);
        let json = std::fs::read_to_string(&path)?;
        let components: ComponentIds = serde_json::from_str(&json)?;
        Ok(Self {
            filename,
            components,
        })
    }

    fn filename(network: &DfxWrapperNetwork) -> String {
        let prefix = "component_ids";
        match network {
            DfxWrapperNetwork::IC => format!("{}_ic.json", prefix),
            _ => format!(
                "{}_{}.json",
                prefix,
                url_to_filename_for_dfx_local(&network.value())
            ),
        }
    }

    pub fn save(&self, dir_path: &str) -> anyhow::Result<()> {
        let path = format!("{}/{}", dir_path, self.filename);
        let json = serde_json::to_string_pretty(&self.components)?;
        std::fs::write(&path, json)?;
        Ok(())
    }

    pub fn add(&mut self, name: String, id: String) {
        self.components.insert(name, id);
    }

    pub fn get(&self, name: &str) -> Option<String> {
        self.components.get(name).cloned()
    }

    pub fn contains_key(&self, name: &str) -> bool {
        self.components.contains_key(name)
    }

    pub fn get_all_entries(&self) -> Vec<(String, String)> {
        self.components
            .iter()
            .map(|(k, v)| (k.clone(), v.clone()))
            .collect()
    }
}

fn url_to_filename_for_dfx_local(url: &str) -> String {
    url.replace(":", "_").replace(".", "_").replace("/", "_")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_url_to_filename_for_dfx_local() {
        assert_eq!(
            url_to_filename_for_dfx_local("http://localhost:8000"),
            "http___localhost_8000"
        );
        assert_eq!(
            url_to_filename_for_dfx_local("http://127.0.0.1:4943/"),
            "http___127_0_0_1_4943_"
        );
    }
}
