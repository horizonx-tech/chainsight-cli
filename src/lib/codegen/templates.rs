use crate::lib::utils::paths;

pub fn root_cargo_toml(
    component_ids: Vec<String>,
    with_bindings: bool,
    with_accessors: bool,
) -> String {
    let mut members = component_ids
        .iter()
        .flat_map(|x| vec![format!("canisters/{}", x), format!("logics/{}", x)])
        .collect::<Vec<String>>();
    if with_bindings {
        members.push("bindings/*".to_string());
    }
    if with_accessors {
        members.push("accessors/*".to_string());
    }

    format!(
        r#"[workspace]
members = [{}]

[workspace.package]
version = "0.1.0"
edition = "2021"

[workspace.dependencies]
candid = "0.9.6"
ic-cdk = "0.11.3"
ic-cdk-macros = "0.8.1"
ic-cdk-timers = "0.5.0"
ic-stable-structures = "0.5.5"
serde = "1.0.163"
serde_bytes = "0.11.12"
serde_json = "1.0.108"
hex = "0.4.3"

ic-web3-rs = "0.1.4"
ic-solidity-bindgen = "0.1.11"
chainsight-cdk-macros = {{ git = "https://github.com/horizonx-tech/chainsight-sdk.git", rev= "011a2b066743870ccc3b205f34fbcb028e11e880" }}
chainsight-cdk = {{ git = "https://github.com/horizonx-tech/chainsight-sdk.git", rev= "011a2b066743870ccc3b205f34fbcb028e11e880" }}
"#,
        members
            .iter()
            .map(|x| format!(r#""{}""#, x))
            .collect::<Vec<String>>()
            .join(", ")
    )
}

pub fn logic_cargo_toml(
    project_name: &str,
    with_bindings: bool,
    dependencies: Vec<String>,
) -> String {
    let txt = format!(
        r#"[package]
name = "{}"
version.workspace = true
edition.workspace = true

[lib]
crate-type = ["rlib"]

[dependencies]
candid.workspace = true
ic-cdk.workspace = true
ic-cdk-macros.workspace = true
ic-cdk-timers.workspace = true
ic-stable-structures.workspace = true
serde.workspace = true
serde_bytes.workspace = true
serde_json.workspace = true
hex.workspace = true

ic-web3-rs.workspace = true
ic-solidity-bindgen.workspace = true
chainsight-cdk-macros.workspace = true
chainsight-cdk.workspace = true

{}
{}
"#,
        project_name,
        if with_bindings {
            paths::bindings_dependency(project_name)
        } else {
            "".to_string()
        },
        if dependencies.is_empty() {
            "".to_string()
        } else {
            paths::accessors_dependency(project_name)
        }
    );

    txt
}

pub fn accessors_cargo_toml(project_name: &str, dependencies: Vec<String>) -> String {
    let txt = format!(
        r#"[package]
name = "{}"
version.workspace = true
edition.workspace = true

[lib]
crate-type = ["rlib"]

[dependencies]
candid.workspace = true
ic-cdk.workspace = true
ic-cdk-macros.workspace = true
ic-cdk-timers.workspace = true
ic-stable-structures.workspace = true
serde.workspace = true
serde_bytes.workspace = true
hex.workspace = true

ic-web3-rs.workspace = true
ic-solidity-bindgen.workspace = true
chainsight-cdk-macros.workspace = true
chainsight-cdk.workspace = true

{}
"#,
        paths::accessors_name(project_name),
        dependencies
            .iter()
            .map(|x| paths::bindings_dependency(x))
            .collect::<Vec<String>>()
            .join("\n")
    );

    txt
}

pub fn canister_project_cargo_toml(project_name: &str, with_bindings: bool) -> String {
    format!(
        r#"[package]
name = "{}"
version.workspace = true
edition.workspace = true

[lib]
crate-type = ["cdylib"]

[dependencies]
candid.workspace = true
ic-cdk.workspace = true
ic-cdk-macros.workspace = true
ic-cdk-timers.workspace = true
ic-stable-structures.workspace = true
serde.workspace = true
serde_bytes.workspace = true
hex.workspace = true

ic-web3-rs.workspace = true
ic-solidity-bindgen.workspace = true
chainsight-cdk-macros.workspace = true
chainsight-cdk.workspace = true

ciborium = "0.2.1"

{}
{}
"#,
        paths::canister_name(project_name),
        paths::logic_dependency(project_name),
        if with_bindings {
            paths::bindings_dependency(project_name)
        } else {
            "".to_string()
        }
    )
}

pub fn bindings_cargo_toml(component: &str) -> String {
    format!(
        r#"[package]
name = "{}"
version.workspace = true
edition.workspace = true

[lib]
crate-type = ["rlib"]

[dependencies]
candid.workspace = true
ic-cdk.workspace = true
ic-stable-structures.workspace = true
serde.workspace = true
serde_bytes.workspace = true

chainsight-cdk-macros.workspace = true
chainsight-cdk.workspace = true
"#,
        paths::bindings_name(component)
    )
}

pub fn dfx_json(project_ids: Vec<String>) -> String {
    let canisters = project_ids
        .iter()
        .map(|label| {
            format!(
                r#"    "{}": {{
      "type": "custom",
      "candid": "./{}.did",
      "wasm": "./{}.wasm",
      "metadata": [
        {{
          "name": "candid:service",
          "visibility": "public"
        }}
      ]
    }}"#,
                label, label, label
            )
        })
        .collect::<Vec<String>>()
        .join(",\n");

    let result = format!(
        r#"{{
  "version": 1,
  "canisters": {{
{}
  }},
  "defaults": {{
    "build": {{
      "packtool": "",
      "args": ""
    }}
  }},
  "output_env_file": ".env"
}}
"#,
        canisters
    );

    result
}

pub fn gitignore() -> String {
    r#"src/__interfaces
src/accessors
src/bindings
src/canisters
src/target
/artifacts/*
!/artifacts/dfx.json
!/artifacts/canister_ids.json
.env
"#
    .to_string()
}

#[cfg(test)]
mod tests {
    use insta::assert_display_snapshot;

    use super::*;

    const PROJECT_NAME: &str = "sample";

    #[test]
    fn test_snapshot_root_cargo_toml() {
        let project_ids = vec![
            "sample_snapshot".to_string(),
            "sample_lens".to_string(),
            "sample_relayer".to_string(),
        ];
        assert_display_snapshot!(
            "snapshot_root_cargo_toml",
            root_cargo_toml(project_ids.clone(), false, false)
        );
        assert_display_snapshot!(
            "snapshot_root_cargo_toml_with_bindings",
            root_cargo_toml(project_ids.clone(), true, false)
        );
        assert_display_snapshot!(
            "snapshot_root_cargo_toml_with_accessors",
            root_cargo_toml(project_ids.clone(), true, true)
        );
    }

    #[test]
    fn test_snapshot_logic_cargo_toml() {
        let dependencies = vec!["sample_snapshot".to_string(), "sample_lens".to_string()];
        assert_display_snapshot!(logic_cargo_toml(PROJECT_NAME, false, dependencies))
    }

    #[test]
    fn test_snapshot_accessors_cargo_toml() {
        let dependencies = vec!["sample_snapshot".to_string(), "sample_lens".to_string()];
        assert_display_snapshot!(accessors_cargo_toml(PROJECT_NAME, dependencies))
    }

    #[test]
    fn test_snapshot_canister_cargo_toml() {
        assert_display_snapshot!(canister_project_cargo_toml(PROJECT_NAME, false))
    }

    #[test]
    fn test_snapshot_bindings_cargo_toml() {
        assert_display_snapshot!(bindings_cargo_toml(PROJECT_NAME))
    }

    #[test]
    fn test_snapshot_dfx_json() {
        let project_ids = vec![
            "sample_snapshot".to_string(),
            "sample_lens".to_string(),
            "sample_relayer".to_string(),
        ];
        assert_display_snapshot!(dfx_json(project_ids))
    }

    #[test]
    fn test_snapshot_gitignore() {
        assert_display_snapshot!(gitignore())
    }
}
