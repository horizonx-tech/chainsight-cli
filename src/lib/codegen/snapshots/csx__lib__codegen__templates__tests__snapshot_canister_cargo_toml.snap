---
source: src/lib/codegen/templates.rs
expression: canister_project_cargo_toml(PROJECT_NAME)
---
[package]
name = "sample_canister"
version.workspace = true
edition.workspace = true

[lib]
crate-type = ["cdylib"]

[dependencies]
candid.workspace = true
ic-cdk.workspace = true
ic-cdk-macros.workspace = true
ic-cdk-timers.workspace = true
ic-stable-structures.workspace = true
serde.workspace = true
serde_bytes.workspace = true
hex.workspace = true

ic-web3-rs.workspace = true
ic-solidity-bindgen.workspace = true
chainsight-cdk-macros.workspace = true
chainsight-cdk.workspace = true

sample = { path = "../../logics/sample" }

